spec/app/domain/authentication/authn_iam/authenticator_spec.rb

```diff
@@ -3,103 +3,186 @@
+            'CONJ00063E Verification of IAM identity failed with exception: Timeout::Error'
+          )
+      end
+    end
   end
-
-  it "fails invalid login with AWS IAM role (based on AWS response)" do
-    subject = authenticator_instance
-    identity_hash = subject.identity_hash(valid_response)
-
-    expect(subject.iam_role_matches?(invalid_login, identity_hash)).to eq(false)
+  context 'for different roles' do
+    context 'when login does not incluce `host`' do
+      let(:payload) do
+        double('AuthenticationParameters', username: 'foo/bar/baz/188945769008/conjur-role', credentials: valid_global_headers)
+      end
+      it 'succeeds', vcr: 'authenticators/authn-iam/valid-global-headers' do
+        expect(authenticator.valid?(payload)).to be(true)
+      end
+      context 'when login is in the root policy' do
+        let(:payload) do
+          double('AuthenticationParameters', username: '188945769008/conjur-role', credentials: valid_global_headers)
+        end
+        it 'fails', vcr: 'authenticators/authn-iam/valid-global-headers' do
+          expect(authenticator.valid?(payload)).to be(false)
+        end
+      end
+    end
+    context 'when login starts with `host`' do
+      context 'when role has multiple policy nestings' do
+        let(:payload) do
+          double('AuthenticationParameters', username: 'host/foo/bar/baz/188945769008/conjur-role', credentials: valid_global_headers)
+        end
+        it 'succeeds', vcr: 'authenticators/authn-iam/valid-global-headers' do
+          expect(authenticator.valid?(payload)).to be(true)
+        end
+      end
+      context 'when account-id is missing from username' do
+        let(:payload) do
+          double('AuthenticationParameters', username: 'host/foo/bar/baz/conjur-role', credentials: valid_global_headers)
+        end
+        it 'fails', vcr: 'authenticators/authn-iam/valid-global-headers' do
+          expect(authenticator.valid?(payload)).to be(false)
+        end
+      end
+      context 'when login has not host  not include a resource id' do
+        let(:payload) do
+          double('AuthenticationParameters', username: 'host/foo/188945769008', credentials: valid_global_headers)
+        end
+        it 'fails', vcr: 'authenticators/authn-iam/valid-global-headers' do
+          expect(authenticator.valid?(payload)).to be(false)
+        end
+      end
+      context 'when login is in the root policy' do
+        let(:payload) do
+          double('AuthenticationParameters', username: 'host/188945769008/conjur-role', credentials: valid_global_headers)
+        end
+        it 'succeeds', vcr: 'authenticators/authn-iam/valid-global-headers' do
+          expect(authenticator.valid?(payload)).to be(true)
+        end
+      end
+      context 'when login does not include an AWS resource id' do
+        let(:payload) do
+          double('AuthenticationParameters', username: 'host/aws/188945769008', credentials: valid_global_headers)
```

Comment:
Just curious - is this test unique from the one beginning on line 162 (missing resource ID)? It's not clear to me why the logic would differ between these two usernames: `host/aws/188945769008` vs. `host/foo/188945769008`